File: develop/counter.c
/*
 * counter.c - YSD8800 ISA2.2 カウンタサンプル (C)
 * scc22 v1.00
 *
 * 0〜9 を1行ずつ出力して "Done!\n" で終了する。
 *
 * ビルド: make counter_c
 */

void putchar(int c);

/* 改行出力 */
void newline() { putchar(10); }

/* 1桁の数字を出力 */
void put_digit(int n) {
    putchar(n + 48);   /* '0' = 48 */
}

/* カウンタメイン */
int main() {
    int i;
    i = 0;
    while (i < 10) {
        put_digit(i);
        newline();
        i = i + 1;
    }
    /* Done! */
    putchar('D'); putchar('o'); putchar('n'); putchar('e'); putchar('!');
    newline();
    return 0;
}

File: src/toolchain/samples/counter.fs
\ counter.fs - YSD8800 ISA2.2 カウンタサンプル (Forth)
\ Force v1.0
\
\ 0〜9 を1行ずつ出力して "Done!" で終了する。
\
\ ビルド: make counter_fs

$FC80 CONSTANT UART-TX
$FC84 CONSTANT UART-STAT

: emit-char  ( c -- )
    BEGIN UART-STAT @ 0<> UNTIL
    UART-TX ! ;

: emit-nl  ( -- )  10 emit-char ;

: put-digit  ( n -- )
    48 +  emit-char ;   \ '0' = 48

: main  ( -- )
    0
    BEGIN
        DUP 10 <
    WHILE
        DUP put-digit
        emit-nl
        1+
    REPEAT
    DROP
    \ Done!
    68 emit-char   \ D
    111 emit-char  \ o
    110 emit-char  \ n
    101 emit-char  \ e
    33 emit-char   \ !
    emit-nl
    HALT ;
